File: scripts/fetchDepositorsTelos.js


const { ethers } = require("ethers");
const fs         = require("fs");
const path       = require("path");

// ------------------------------------------
// 1. Configuration
// ------------------------------------------
const RPC_URLS = ["https://rpc.telos.net"];

const CONTRACTS = {
  lendingPoolAddressProvider: "0x703cF2C85EA76C54bd863337585673B3DF8FCE72",
  protocolDataProvider:       "0x6DE58d6dBECF87D7cE972f6E4838fEeCc63B4c5e",
};

const USDM_UNDERLYING = ethers.utils.getAddress("0x8f7D64ea96D729EF24a0F30b4526D47b80d877B9");
const USDT_UNDERLYING = ethers.utils.getAddress("0x975Ed13fa16857E83e7C493C7741D556eaaD4A3f");

const PADDED_USDM_TOPIC = ethers.utils.hexZeroPad(USDM_UNDERLYING, 32);
const PADDED_USDT_TOPIC = ethers.utils.hexZeroPad(USDT_UNDERLYING, 32);
// (we define these in case you later want to filter by reserve)
const TARGET_TOPICS = [PADDED_USDM_TOPIC, PADDED_USDT_TOPIC];

const BLOCK_INCREMENT = 100_000;
const BALANCE_BLOCK   = 416_907_698;

const OUT_DIR        = path.resolve(__dirname, "../json");
const OUT_DEPOSITORS = path.join(OUT_DIR, "depositors_usdm_usdt_taraxa.json");

// ------------------------------------------
// 2. Load your single-ABI file
// ------------------------------------------
const lendingPoolAbi = require("../config/abis.json");  // must be a top-level array

// ------------------------------------------
// 3. Helpers
// ------------------------------------------
async function getProvider(rpcs) {
  for (const rpc of rpcs) {
    try {
      const provider = new ethers.providers.JsonRpcProvider(rpc);
      await provider.getBlockNumber();
      console.log(`✅ Connected to RPC: ${rpc}`);
      return provider;
    } catch (e) {
      console.error(`❌ RPC failed (${rpc}): ${e.message}`);
    }
  }
  throw new Error("All RPC endpoints failed.");
}

async function fetchDepositorsInRange(provider, contractAddress, fromBlock, toBlock) {
  const depositSig   = "Deposit(address,address,address,uint256,uint16)";
  const depositTopic = ethers.utils.id(depositSig);
  const allDepositors = new Set();

  for (let start = fromBlock; start <= toBlock; start += BLOCK_INCREMENT) {
    const end = Math.min(start + BLOCK_INCREMENT - 1, toBlock);
    console.log(`🔍 Scanning Deposit events from ${start} to ${end}…`);

    try {
      const logs = await provider.getLogs({
        fromBlock: start,
        toBlock:   end,
        address:   contractAddress,
        topics:    [[depositTopic]],
        // if you wanted to filter by reserve (USDM/USDT), you could instead:
        // topics: [ [depositTopic], null, TARGET_TOPICS ]
      });
      console.log(`   → ${logs.length} log(s) found`);

      for (const log of logs) {
        if (log.topics.length > 1) {
          const raw      = log.topics[1];  // indexed depositor
          const depositor = ethers.utils.getAddress("0x" + raw.slice(-40));
          allDepositors.add(depositor);
        }
      }
    } catch (e) {
      console.error(`   ! Error fetching ${start}–${end}: ${e.message}`);
    }
  }

  return Array.from(allDepositors);
}

// ------------------------------------------
// 4. Main Workflow
// ------------------------------------------
async function main() {
  const provider = await getProvider(RPC_URLS);

  // Instantiate the AddressProvider contract
  const lpProvider = new ethers.Contract(
    CONTRACTS.lendingPoolAddressProvider,
    lendingPoolAbi,
    provider
  );

  // Resolve the active LendingPool address
  const lendingPoolAddr = await lpProvider.getLendingPool();
  console.log(`🏦 LendingPool address resolved to: ${lendingPoolAddr}\n`);

  // Fetch depositors
  console.log(`🚀 Scanning deposit events up to block ${BALANCE_BLOCK}…`);
  const depositors = await fetchDepositorsInRange(
    provider,
    lendingPoolAddr,
    413_996_255,
    BALANCE_BLOCK
  );
  console.log(`\n🎉 Found ${depositors.length} unique depositor(s).`);

  // Write results
  fs.mkdirSync(OUT_DIR, { recursive: true });
  fs.writeFileSync(OUT_DEPOSITORS, JSON.stringify(depositors, null, 2));
  console.log(`💾 Depositors saved to ${OUT_DEPOSITORS}`);
}

main().catch(err => {
  console.error(err);
  process.exit(1);
});

File: taraxa_lending/fetchDepositorsTelos.py
import json
import os
import sys
from hexbytes import HexBytes
from web3 import Web3, HTTPProvider
import config.abis as abis

# ------------------------------------------
# 1. Configuration
# ------------------------------------------
RPC_URLS = ["https://rpc-private.mainnet.taraxa.io"]

CONTRACTS = {
    "lendingPoolAddressProvider": "0x0EdbA5d821B9BCc1654aEf00F65188de636951fa",
    "protocolDataProvider":       "0x0208E7B745591f6c2F02B4DcF53B3e1f11c671df",
}


BLOCK_INCREMENT   = 1000
BALANCE_BLOCK     = 19916232  # <-- scan stops here

OUT_DEPOSITORS    = "json/depositors_usdm_usdt_taraxa.json"
OUT_BALANCES      = f"json/depositor_balances_block_{BALANCE_BLOCK}.json"

# ------------------------------------------
# 2. Helpers
# ------------------------------------------
def get_provider(rpcs):
    for rpc in rpcs:
        try:
            w3 = Web3(HTTPProvider(rpc))
            _ = w3.eth.block_number
            print(f"Successfully connected to RPC: {rpc}")
            return w3
        except Exception as e:
            print(f"RPC failed ({rpc}): {e}", file=sys.stderr)
    raise RuntimeError("All RPC endpoints failed.")

# ------------------------------------------
# 3. Fetch and Print Deposit Events in Chunks
# ------------------------------------------
def fetch_depositors_in_range(w3, contract_address, from_block, to_block):
    """
    Fetch all Deposit events between from_block and to_block in BLOCK_INCREMENT chunks.
    Returns a set of depositor addresses.
    """
    # Only one signature → one topic
    deposit_signature = "Deposit(address,address,address,uint256,uint16)"
    deposit_topic = w3.keccak(text=deposit_signature).hex()
    all_depositors = set()

    for start in range(from_block, to_block + 1, BLOCK_INCREMENT):
        end = min(start + BLOCK_INCREMENT - 1, to_block)
        print(f"Scanning Deposit events from {start} to {end}…")

        filter_params = {
            "fromBlock": start,
            "toBlock":   end,
            "address":   contract_address,
            "topics":    [[deposit_topic]],
        }

        try:
            logs = w3.eth.get_logs(filter_params)
            print(f"  → {len(logs)} logs")

            for log in logs:
                # topic[1] is the indexed depositor address
                if len(log["topics"]) > 1:
                    addr_hex = log["topics"][1].hex()  # "0x0000…abcd"
                    depositor = Web3.to_checksum_address("0x" + addr_hex[-40:])
                    all_depositors.add(depositor)

        except Exception as e:
            print(f"Error fetching {start}–{end}: {e}", file=sys.stderr)

    return all_depositors

# ------------------------------------------
# 4. Main Workflow
# ------------------------------------------
def main():
    w3 = get_provider(RPC_URLS)

    # Resolve LendingPool
    lp_provider = w3.eth.contract(
        address=CONTRACTS["lendingPoolAddressProvider"],
        abi=abis.lendingPoolAddressProvider()
    )
    lending_pool_addr = lp_provider.functions.getLendingPool().call()
    print(f"LendingPool address resolved to: {lending_pool_addr}")

    # 4a) Scan depositors up to BALANCE_BLOCK
    print(f"\nScanning deposit events up to block {BALANCE_BLOCK}...")
    depositors = fetch_depositors_in_range(w3, lending_pool_addr,16792600, BALANCE_BLOCK)


if __name__ == "__main__":
    main()